File: lib/progressbar/progressbar.dart
import "package:angular2/angular2.dart";

const progressConfig = const { "animate" : true, "max" : 100};
// todo: progress element conflict with bootstrap.css

// todo: need hack: replace host element with div
@Directive(selector: "[n2s-progress]",
    inputs: const [ "animate", "max"],
    host: const { "class" : "progress", "[attr.max]" : "max"})
class Progress implements OnInit {
  num _max;

  bool animate;

  List<dynamic> bars = [];

  Progress() {}

  ngOnInit() {
    animate = !identical(animate, false);
    max = max is num ? max : progressConfig['max'];
  }

  num get max => _max;

  set max(num v) {
    _max = v;
    bars.forEach((Bar bar) {
      bar.recalculatePercentage();
    });
  }

  addBar(Bar bar) {
    if (!animate) {
      bar.transition = "none";
    }
    bars.add(bar);
  }

  removeBar(Bar bar) {
    bars.remove(bar);
  }
}
// todo: number pipe

// todo: use query from progress?
@Component (selector: "n2s-bar",
    inputs: const ["type", "value"],
    template: '''
  <div class="progress-bar"
    style="min-width: 0;"
    role="progressbar"
    [ngClass]="type"
    [ngStyle]="{'width': (percent < 100 ? percent : 100).toString() + \'%\', transition: transition}"
    aria-valuemin="0"
    [attr.aria-valuenow]="value"
    [attr.aria-valuetext]="percent.toStringAsFixed(0) + \'%\'"
    [attr.aria-valuemax]="max"
    ><ng-content></ng-content></div>
''',
    directives: const [NgStyle, NgClass],
    encapsulation: ViewEncapsulation.None)
class Bar implements OnInit, OnDestroy {
  Progress progress;

  String _type;

  String get type => _type != null ? 'progress-bar-' + _type : null;
  set type(String type) => _type = type;

  num percent = 0;

  String transition;

  num _value;

  num max;

  Bar(@Host() this.progress) {}

  ngOnInit() {
    progress.addBar(this);
  }

  ngOnDestroy() {
    progress.removeBar(this);
  }

  num get value => _value;

  set value(num v) {
    if (v == null || v == 0) {
      return;
    }
    _value = v;
    recalculatePercentage();
  }

  recalculatePercentage() {
    percent = 100 * value / progress.max;
    var totalPercentage = progress.bars.fold(0, (total, bar) {
      return total + bar.percent;
    });
    if (totalPercentage > 100) {
      percent -= totalPercentage - 100;
    }
  }
}

@Component (selector: "n2s-progressbar",
    inputs: const [ "animate", "max", "type", "value"],
    template: '''
    <div n2s-progress [animate]="animate" [max]="max">
      <n2s-bar [type]="type" [value]="value">
          <ng-content></ng-content>
      </n2s-bar>
    </div>
  ''', directives: const [ Progress, Bar])
class Progressbar {
  bool animate;

  num max;

  String type;

  num value;
}

const List PROGRESSBAR_DIRECTIVES = const [ Progress, Bar, Progressbar];

File: components/progressbar/progressbar.ts
import {
  Component, Directive,
  OnInit, OnDestroy,
  Host, Input, HostBinding
} from 'angular2/core';
import { NgClass, NgStyle } from 'angular2/common';

const progressConfig = {
  animate: true,
  max: 100
};

// todo: progress element conflict with bootstrap.css
// todo: need hack: replace host element with div
@Directive({ selector: 'bs-progress, [progress]' })
export class Progress implements OnInit {
  @Input() public animate:boolean;

  @HostBinding('attr.max')
  @Input() public get max():number {
    return this._max;
  }

  @HostBinding('class') private addClass = 'progress';

  public set max(v:number) {
    this._max = v;
    this.bars.forEach((bar:Bar) => {
      bar.recalculatePercentage();
    });
  }

  public bars:Array<any> = [];

  private _max:number;

  constructor() {
  }

  ngOnInit() {
    this.animate = this.animate !== false;
    this.max = typeof this.max === 'number' ? this.max : progressConfig.max;
  }


  public addBar(bar:Bar) {
    if (!this.animate) {
      bar.transition = 'none';
    }
    this.bars.push(bar);
  }

  public removeBar(bar:Bar) {
    this.bars.splice(this.bars.indexOf(bar), 1);
  }
}

// todo: number pipe
// todo: use query from progress?
@Component({
  selector: 'bar, [bar]',
  directives: [NgClass, NgStyle],
  template: `
  <div class="progress-bar"
    style="min-width: 0;"
    role="progressbar"
    [ngClass]="type && 'progress-bar-' + type"
    [ngStyle]="{width: (percent < 100 ? percent : 100) + '%', transition: transition}"
    aria-valuemin="0"
    [attr.aria-valuenow]="value"
    [attr.aria-valuetext]="percent.toFixed(0) + '%'"
    [attr.aria-valuemax]="max"
    ><ng-content></ng-content></div>
`
})
export class Bar implements OnInit, OnDestroy {
  @Input() public type:string;
  @Input() public get value():number {
    return this._value;
  }

  public set value(v:number) {
    if (!v && v !== 0) {
      return;
    }
    this._value = v;
    this.recalculatePercentage();
  }

  public percent:number = 0;
  public transition:string;

  private _value:number;

  constructor(@Host() public progress:Progress) {
  }

  ngOnInit() {
    this.progress.addBar(this);
  }

  ngOnDestroy() {
    this.progress.removeBar(this);
  }

  public recalculatePercentage() {
    this.percent = +(100 * this.value / this.progress.max).toFixed(2);

    let totalPercentage = this.progress.bars.reduce(function (total, bar) {
      return total + bar.percent;
    }, 0);

    if (totalPercentage > 100) {
      this.percent -= totalPercentage - 100;
    }
  }
}

@Component({
  selector: 'progressbar, [progressbar]',
  directives: [Progress, Bar],
  template: `
    <div progress [animate]="animate" [max]="max">
      <bar [type]="type" [value]="value">
          <ng-content></ng-content>
      </bar>
    </div>
  `
})
export class Progressbar {
  @Input() private animate:boolean;
  @Input() private max:number;
  @Input() private type:string;
  @Input() private value:number;
}

export const PROGRESSBAR_DIRECTIVES:Array<any> = [Progress, Bar, Progressbar];
/**
 * @deprecated use PROGRESSBAR_DIRECTIVES instead
 * @type {Progress|Bar|Progressbar[]}
 */
export const progressbar:Array<any> = [Progress, Bar, Progressbar];
